### PR TITLE
Update `detect_machine.sh` using that from UFS WM

### LINKED ISSUE #2228: Updates to `detect_machine.sh` needed after the UFSWM updated the return value
### What is wrong?

gfs-utils, gsi, and other repos. have a copy of `detect_machine.sh`.  This was copied from `UFSWM` to maintain consistency.  The UFSWM has updated and removed the `COMPILER` addon to the variable `MACHINE_ID` for all platforms.
This needs to be replicated in the above repos.

### What should have happened?

A consistent return of use of `detect_machine.sh` should return the name of the machine w/o `COMPILER` addon.

### What machines are impacted?

All or N/A

### Steps to reproduce

None

### Additional information

None

### Do you have a proposed solution?

Copy `detect_machine.sh` from ufs-weather-model to the above repos.
https://github.com/ufs-community/ufs-weather-model/blob/develop/tests/detect_machine.sh

### PR DESCRIPTION
**Pull request by HenryRWinterbottom**
# Description

This PR addresses issue #2228. The following is accomplished:

- the global-workflow `ush/detect_machine.sh` is replaced by the UFS weather-model `tests/detect_machine.sh` prepared by @BrianCurtis-NOAA 

Resolves #2228 

# Type of change

- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?

CI/CD will test.

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Review by aerorahul — APPROVED** [review]
Looks good. Thanks.

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
Marking as draft to prevent merging until submodules are updated.

**Comment by WalterKolczynski-NOAA** [comment]
Can we get a checklist of the submodules being updated so we can track the progress and know when this can proceed?

**Comment by HenryRWinterbottom** [comment]
@WalterKolczynski-NOAA @aerorahul Had suggested this as well. 

Do you want me to append it to this PR or to the original issue such that we have here rather than floating around on a Google doc or sheet?

**Comment by WalterKolczynski-NOAA** [comment]
@HenryWinterbottom-NOAA Here, since it is blocking this PR. Having it on the issue wouldn't hurt either.

**Comment by HenryRWinterbottom** [comment]
Current (as of Fri Feb  2 15:00:50 UTC 2024) list of global workflow submodules updated to comply with the UFSWM `detect_machine.sh`:

- [GDASApp](https://github.com/noaa-emc/gdasapp)
- [GSI-monitor](https://github.com/noaa-emc/gsi-monitor)
- [GFS-utils](https://github.com/noaa-emc/gfs-utils)
- [GSI](https://github.com/NOAA-EMC/GSI.git)

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful completion of CI tests.

**Pull request merged**
